### DISCUSSION TITLE
Memory errors with detrending + rolling operations

### DISCUSSION THREAD

**Discussion by tlsw231** [pydata/xarray#6280] [Q&A]
I am trying to perform these two operations in succession on a 3d array with dimensions `(t,y,x)`: 
(i) detrend along `t`
(ii) rolling mean along `t`
I get memory related failures if the size along the `t` dimension becomes too large even though I think I have sufficient memory. That is, the progress bars in the dask dashboard show many "erred" processes and the kernel dies midway. Here is some stand-alone code that mimics what I am trying to with comparable array sizes:

```
# Import a bunch of stuff
import sys
import os 

import math
import xarray as xr
import numpy as np
import dask
from dask.distributed import Client
from dask_jobqueue import SLURMCluster

# Launch the cluster: 8 cores, 340 Gb
cluster = SLURMCluster(cores=8, processes=8, memory="340GB",queue="short",
                       walltime="01:00:00")
print(cluster.job_script())
cluster.scale(8)
client = Client(cluster)

# Array size. My actual application has nt ~ 12000 but this code fails even for the value below
(nt,ny,nx)=(2000,489,655)

# Define the data array
dummy=xr.DataArray(data=np.random.randn(nt,ny,nx),dims=['t','y','x'])

# The size in Gb
dummy.nbytes/1e9
5.12472

# Function to detrend
# Source: https://gist.github.com/rabernat/1ea82bb067c3273a6166d1b1f77d490f 
def detrend_dim(da, dim, deg=1):
    # detrend along a single dimension
    p = da.polyfit(dim=dim, deg=deg)
    fit = xr.polyval(da[dim], p.polyfit_coefficients)
    return da - fit

# Detrend
chunks='auto'
field = dummy.chunk(chunks)
dim = 't'
field_detrend = detrend_dim(field,dim,deg=1)

# Rolling mean
half_win = 8;
field_roll = field_detrend.chunk('auto').rolling(t=2*half_win+1,center=True).mean(skipna=True)

# -----------Kernel dies while trying to execute this step----------------
# Slice along the time dimension to throw out the NaN values 
# from the rolling operation and persist the result
out = ( field_roll.chunk('auto').isel(t=slice(half_win,-half_win)) ).persist()

```
Any suggestions on why the above fails? Thanks.

**Comment by dcherian** [comment]
I think you're running into https://github.com/pydata/xarray/issues/5629 . There is a reshape operation that is extremely inefficient. It might work better if your chunk sizes are small along time and large in space.

**Reply by tlsw231** [reply]
Thanks for the response!

For unknown reasons, the MVE I had posted ran as is without any further changes the next day. I don't know if something was wrong with the compute node on the first day. But my original application is still giving me the same issues. I have 8 years of data available 3 hourly.

One year worth data `(time,y,x)= (2920,489,655)`: I just use `chunks=auto` throughout and the detrending + rolling works without too much fuss.

Two years `(5840,489,655)`: I had to use `chunks=dict(time=-1,y=50,x=100)` to get it to work. So if the detrending dimension is completely unchunked the `persist` step goes through smoothly spilling very little (< 100 mb) to disk. In fact, some of the other options where I chunk the time dimension fail.

Three years `(8760,489,655)`: Am unable to get it to work. I have tried unchunking/chunking time, increasing/decreasing the spatial chunks, nothing seems to make any difference.

I am slightly confused now. The discussion you linked to was useful and began with a description of how unchunking along the time dimension was not working. In my case that is exactly what helped initially (2 years worth data) but fails with more data. So I am stumped how to make progress here. Any suggestions are most welcome.

Output of `xr.show_versions()`:
```
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1160.53.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C
LOCALE: (None, None)
libhdf5: 1.12.1
libnetcdf: 4.8.1

xarray: 0.18.2
pandas: 1.2.5
numpy: 1.21.0
scipy: 1.7.0
netCDF4: 1.5.8
pydap: installed
h5netcdf: None
h5py: None
Nio: None
zarr: 2.8.3
cftime: 1.5.0
nc_time_axis: 1.3.1
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2021.07.0
distributed: 2021.07.0
matplotlib: 3.4.2
cartopy: 0.19.0.post1
seaborn: None
numbagg: None
pint: None
setuptools: 52.0.0.post20210125
pip: 21.1.3
conda: None
pytest: None
IPython: 7.22.0
sphinx: None
```

**Reply by tlsw231** [reply]
I am trying to use the `reshape_block` routine provided in the comments [here](https://github.com/pydata/xarray/issues/5629). The slight difference is that I am using an xarray dataarray as opposed to a dask array (as in the example provided). 

```python
arr = xr.DataArray(np.random.random([100,2,2]),dims=['t','y','x'])
arr = arr.chunk(dict(t=-1,x=1,y=1))

# To shift the time dimension to the end
arr = arr.transpose()

# This is from the comments section in the thread linked to above.
# assumes we're collapsing axes (0,1) and preserving axis 2
output_chunks = (tuple(map(np.product, product(*arr.chunks[:2]))),) + (arr.chunks[-1],)
reshaped = reshape_block(arr, output_chunks)
actual = reshape_block(reshaped, arr.chunks)
assert_eq(actual, arr) # True
```
I am getting the error message:
```
'DataArray' object has no attribute '_meta'
```
triggered by this line in `reshape_block`:
```
res = Array(graph, name, output_chunks, arr.dtype, meta=arr._meta)
```

Any suggestion on how I might fix this? Thanks.

**Reply by mathause** [reply]
You may have to do `arr.data = reshape_block(arr.data, output_chunks)` (untested)

**Reply by tlsw231** [reply]
Thanks! adding `.data` inside the argument to `reshape_block` removed this error.

**Reply by tlsw231** [reply]
I tried both the `apply_ufunc` and `reshape_block` options in the [comment thread](https://github.com/pydata/xarray/issues/5629) but seem to be having the same problem with my original application. This is not a MVE but the essence of what I am doing is something like this:
``` 
1.Read in the field from multiple files using open_mfdataset

2. chunks = <Decide how t, y and x are to be chunked for the detrending step>

3. Chunk field using the chunks in #2 and detrend (either using polyfit or reshape_block + new_polyfit)

4. chunks = <Decide chunk sizes for detrended field before rolling mean step>

5. Compute rolling mean along time using chunking in #4. 
```
Here are links to movies of the dask dashboard for three cases:

 (i) [2 years data](https://www.dropbox.com/s/zfwv6ijrdkv0hqk/works.mov?dl=0): **works** with time (t) unchunked
Chunking in step 2: time unchunked, chunk size ~ 110 mb
Chunking in step 4: Just `'auto'` works

 (ii) [3 years](https://www.dropbox.com/s/ltmv2vp44i2z6qw/fails_time_chunked.mov?dl=0): **fails** with time chunked
Chunking in step 2: time chunked, x and y unchunked, chunk size ~ 50 mb
Chunking in step 4: Neither `'auto'` or chunking along time works

 (iii) [3 years](https://www.dropbox.com/s/kuwpdy6jn7n15me/fails_time_unchunked.mov?dl=0): **fails** with time unchunked
Chunking in step 2: time unchunked, x and y chunked, chunk size ~ 80 mb
Chunking in step 4: Neither `'auto'` or chunking along time works

For cases (ii) and (iii) it is unclear to me what is a good chunking strategy for step 2. Whether I chunk or unchunk the detrended dimension, the result is very similar, as the dask dashboard reveals. 

[PS: I also tried the `apply_ufunc` option in #5933, to no avail as it fails in a nearly identical manner.]

**Reply by dcherian** [reply]
One thing that jumps out is the repeated rechunking. How are your data stored on disk? You should be able to pass the chunking in (2) as an argument to  `open_mfdataset`.

The detrending step (3) does not require a rechunk if you use that `reshape_block` trick. `polyfit` has a `skipna` kwarg  that might help.

rechunking in step (4) should also be unnecessary.

To choose your initial chunksizes, note that rolling along a chunked dimension  uses `map_overlap`. For small chunksizes, this adds a lot of tasks and communication (values have to be transferred between neighbouring blocks). So choose big chunksizes along time in the open_mfdataset call if you can (this should help with the detrending too).

**Reply by tlsw231** [reply]
The data do not appear to be stored on disk with any defined chunk size. When I run `ncdump` I do not see any setting for `chunksizes`. 

I am having trouble forcing xarray to accept the pre-defined `chunks` directly in the `open_mfdataset` call as follows (the data is from the ROMS ocean model):

```python
chunks = dict(ocean_time=None, eta_rho=50, xi_rho=50)

# I have a pre-defined list of variables to be dropped (varlist)
ds = xr.open_mfdataset(
    files,
    chunks=chunks,
    data_vars="minimal",
    coords="minimal",
    compat="override",
    drop_variables=varlist,
)

# Define time range for slicing (3 years)
t1 = int(np.where(ds.ocean_time == np.datetime64("2015-12-25T01:30:00"))[0])
t2 = int(np.where(ds.ocean_time == np.datetime64("2018-12-25T10:30:00"))[0])

# Slice data in time and pick only the values at the topmost model level (s_w = -1, s_rho = -1)
ds = ds.isel(dict(s_rho=-1, s_w=-1, ocean_time=slice(t1, t2 + 1))).copy()

# Drop some unnecessary variables, not sure why they were not dropped in the open_mfdataset call
ds = ds.drop_vars(["Hsbl", "s_w", "s_rho"])
```

When I run the above cell, I get this warning:
`PerformanceWarning: Increasing number of chunks by factor of 10`

I then tried adding the following setting at the top to force dask to accept large chunks and repeated the `open_mfdataset` step. I get the same performance warning.
`dask.config.set({"array.slicing.split_large_chunks": False})`

It does not look like xarray is respecting the chunk size specified in the `open_mfdataset` call, as seen in the image below. Even though I specified the `ocean_time` dimension should not be chunked, it nevertheless is. I am not even sure I understand the chunking calculation here. For the given chunk sizes, why is the number of chunks so large?

![Screen Shot 2022-02-24 at 1 29 51 PM](https://user-images.githubusercontent.com/90293171/155594628-569ae02e-3021-4f0c-9852-508763802b60.png)

**Reply by tlsw231** [reply]
Forgot to mention I have been using `skipna=True` all along while calling `polyfit`.

**Reply by dcherian** [reply]
One file per timestep is really not conducive to what you're doing. I think you'll just have to rechunk with `rechunker` and save a second copy of your dataset.

**Reply by tlsw231** [reply]
Thanks, I will look at the `rechunker` option. Each file has 8 timesteps by the way.

**Reply by dcherian** [reply]
Ah one thing I spot in your dask array image is that the last chunk is giant. I think you have a lot of tiny tiny chunks and then the biggest one is a 25MB. Something has gone wrong with "alignment" of coordinate locations along the time dimension.

**Reply by tlsw231** [reply]
So I came across this [thread](https://stackoverflow.com/questions/56032626/problems-with-chunksize-dask-xarray-zarr) which says the `chunks` keyword in `open_mfdataset` does not work as intended if `concat_dim` is also specified. The solution according to that thread is to add `.chunk(chunks)` at the end. I tried that and the chunks look reasonable after the `open_mfdataset` step but that still not improve anything downstream.

**Reply by tlsw231** [reply]
Regarding the `rechunker` option, I am trying to use the following code (have used this successfully before in a different problem). But I am getting an error in the `to_zarr` step. It appears that even though `sst` is obtained by slicing only the top most level from the original four-dimension field (time,depth,y,x) and should therefore be three-dimensional, the error message indicates it is trying to write out the four-dimensional variable (and runs out of memory). The error message is below the code.
```
prefix = 'sst'
rechunker_maxmem = '2.0GB'

# ---THIS STEP IS GIVING ME ERRORS----------
sst.chunk('auto').to_dataset().to_zarr(savedir + prefix + '.zarr',mode='w')

source_group=zarr.open(savedir + prefix + '.zarr')

# No chunking along detrending dimension (n1 = -1) and chunking along other two dimensions
n1=-1; n2=50; n3=50;

target_chunks = {'temp':{'ocean_time':n1,'eta_rho':n2,'xi_rho':n3},
                      'ocean_time':None,
                      'eta_rho' :None,
                      'xi_rho':None}
    
max_mem = rechunker_maxmem
                       
target_store = savedir + prefix + '_rechunked.zarr'
temp_store   = savedir + prefix + '_rechunked-tmp.zarr'

array_plan = rechunk(source_group, target_chunks, max_mem, target_store, 
                             temp_store=temp_store)
array_plan

with ProgressBar():
    array_plan.execute()
    
sst_rechunk=xr.open_zarr(savedir + prefix + '_rechunked.zarr', chunks='auto')
```
This is the error message:
```
MemoryError: Unable to allocate 354. GiB for an array with shape (5928, 50, 489, 655) and data type float32
```

It is trying to write out a 4d array even though the field is actually 3d:
![Screen Shot 2022-02-25 at 2 21 27 PM](https://user-images.githubusercontent.com/90293171/155800224-d2a42d3d-fd24-4605-815f-deeffbd5e56f.png)

**Reply by dcherian** [reply]
I looked in to this a bit: https://github.com/NCAR/esds/blob/fcc6909289b8aa4931ff4d2a996c6d0a64ed2f5a/posts/2022/dask-debug-detrend-1.ipynb

Unfortunately I think the problem is a bad dask regression: https://github.com/dask/dask/issues/6916

**Reply by dcherian** [reply]
The dask bug is now fixed and things seem to work OK (https://ncar-esds--166.org.readthedocs.build/en/166/posts/2022/dask-debug-detrend-1.html) That new polyval + that blockwise reshape trick should work decently. @tlsw231 this could be a good time to retry it.

**Reply by tlsw231** [reply]
Thanks so much for pursuing this! I read your earlier message too, apologies for not responding sooner. The link in your latest message is not working for me. 

Even without the new `reshape` routine, I am able to process my data after having made some changes (my dataset is 3-hourly output for 8 years). [I am having some issues with the `reshape` routine for my dataset even though it works perfectly well for the test cases I created. The routine itself works but the detrended field is having ridiculously large magnitudes like 10^15 etc. But this is something for me to sort out.]. Here is what I am doing presently:

1.  Resample everything to daily-averaged values and work with those.
2. I found that using `isel` to slice in space and time simultaneously was giving me large uneven chunks in `time`, like what you also see in your notebook that you shared. I found that using `isel` twice, first on spatial coordinates and then again on the `time` coordinate removed this problem. The latter option gives me even chunks along all axes without one very large chunk, as was happening earlier. 
3. Use the setting below at the start of my notebook, am not sure if this always works as intended.
  `dask.config.set(**{'array.slicing.split_large_chunks': True})` 

If I can sort out my problem with the `reshape` routine perhaps I can even work directly with the 3-hour data.

**Reply by dcherian** [reply]
It's here now: https://ncar.github.io/esds/posts/2022/dask-debug-detrend/

**Reply by tlsw231** [reply]
I hope I am not violating any policies wrt posting here but I had also raised a related [query](https://github.com/pydata/xarray/discussions/6322) on trying to use `apply_ufunc` for the detrending. Is there any reason to expect the `apply_ufunc` option might be faster? [I could not get `apply_ufunc` to work if I progressively increased the data size, as described in that thread.]
